### PR TITLE
Bug fix fetch_callback of Partial Subset DAG

### PR DESCRIPTION
**Pull request by sungwy**
I believe that there may be a bug where if a partial_subset dag is generated on a DAG with `on_success_callback` or `on_failure_callback` functions defined on it, then it might try to look for a TI that is not actually a part of the partial subset DAG. This is because we fetch the last TI available from the DagRun within `DAG.fetch_callback`, and a DagRun isn't aware of the concept of a partial subset. Hence `ti.task = dag.get_task(ti.task_id)` results in a `TaskNotFound` exception.

In this PR, we are:

1. Adding DAG callback functions to the test DAG defined in  tests/jobs/test_backfill_job.py::TestBackfillJob::test_sub_set_subdag to cover this test case (which fails without below patch)
2. If dag.partial, we filter out tis where ti.state == State.NONE within `DAG.fetch_callback`

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/models/dag.py [hunk 1]
```diff
@@ -1428,6 +1428,13 @@ def fetch_callback(
             dagrun = DAG.fetch_dagrun(dag_id=dag.dag_id, run_id=dag_run_id, session=session)
             callbacks = callbacks if isinstance(callbacks, list) else [callbacks]
             tis = dagrun.get_task_instances(session=session)
+            # tis from a dagrun may not be a part of dag.partial_subset,
+            # since dag.partial_subset is a subset of the dag.
+            # This ensures that we will only use the accessible TI
+            # context for the callback.
+            if dag.partial:
+                tis = [ti for ti in tis if not ti.state == State.NONE]
+            # tis = [ti for ti in tis if ti.task_id in dag.task_dict.keys()]
```
Comment (Lee-W) [inline airflow/models/dag.py]: Should we remove this comment?

Comment (sungwy) [reply airflow/models/dag.py]: Thanks for the catch @Lee-W

**Review by vincbeck — APPROVED** [review]

**Comment by syun64** [comment]
Thank you for the review @vincbeck . Would you be able to merge this in? Unfortunately I do not have write access to the repository

**Comment by vincbeck** [comment]
Done! Thanks for the PR!

**Pull request merged**